core/Pos/Util/Util.hs

```diff
@@ -260,6 +261,10 @@ type instance ChannelT (Ether.TaggedTrans tag t m) = ChannelT m
 maybeThrow :: (MonadThrow m, Exception e) => e -> Maybe a -> m a
 maybeThrow e = maybe (throwM e) pure
 
+-- | Fail or return result depending on what is stored in 'Either'.
+eitherToFail :: (MonadFail m, ToString s) => Either s a -> m a
```

Comment:
> MonadError Text m ~ MonadFail m

Not true. `MonadError` also gives you the `catch` operation. I'd avoid `MonadFail` whenever possible.

Instead of `MonadError` I recommend creating a type per exception and using them with `Ether.MonadExcept'`. One disadvantage is that each layer of `ExceptT` adds a bit of performance overhead, but I can implement flattening for `Ether.MonadExcept` like I did for `Ether.MonadReader`, so don't worry about this.

If you don't want to rewrite error handling just yet, I recommend using `MonadError`.